apiserver/facades/client/modelmanager/modelmanager.go

```diff
@@ -1614,15 +1668,105 @@ func (m *ModelManagerAPI) ChangeModelCredential(args params.ChangeModelCredentia
 			results[i].Error = common.ServerError(err)
 		}
 	}
-	return params.ErrorResults{results}, nil
+	return params.ErrorResults{Results: results}, nil
+}
+
+// ValidateModelUpgrades validates if a model is allowed to perform an upgrade.
+// Examples of why you would want to block a model upgrade, would be situations
+// like upgrade-series. If performing an upgrade-series we don't know the
+// current status of the machine, so performing an upgrade-model can lead to
+// bad unintended errors down the line.
+func (m *ModelManagerAPI) ValidateModelUpgrades(args params.ValidateModelUpgradeParams) (params.ErrorResults, error) {
+	if err := m.check.ChangeAllowed(); err != nil {
+		return params.ErrorResults{}, errors.Trace(err)
+	}
+
+	controllerAdmin, err := m.authorizer.HasPermission(permission.SuperuserAccess, m.state.ControllerTag())
+	if err != nil {
+		return params.ErrorResults{}, errors.Trace(err)
+	}
+	// Only controller or model admin can change cloud credential on a model.
+	checkModelAccess := func(tag names.ModelTag) error {
+		if controllerAdmin {
+			return nil
+		}
+		modelAdmin, err := m.authorizer.HasPermission(permission.AdminAccess, tag)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if modelAdmin {
+			return nil
+		}
+		return common.ErrPerm
+	}
+
+	results := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(args.Models)),
+	}
+	for i, arg := range args.Models {
+		modelTag, err := names.ParseModelTag(arg.ModelTag)
+		if err != nil {
+			results.Results[i] = params.ErrorResult{Error: common.ServerError(err)}
+			continue
+		}
+		if err := checkModelAccess(modelTag); err != nil {
+			results.Results[i] = params.ErrorResult{Error: common.ServerError(err)}
+			continue
+		}
+
+		// We now need to access the state pool for that given model.
+		st, err := m.statePool.Get(modelTag.Id())
+		if err != nil {
+			results.Results[i] = params.ErrorResult{Error: common.ServerError(err)}
+			continue
+		}
+		defer st.Release()
+
+		model, err := st.Model()
+		if err != nil {
+			results.Results[i] = params.ErrorResult{Error: common.ServerError(err)}
+			continue
+		}
+
+		// If it's a controller model then we don't require certain validation
+		// checks.
+		if model.IsControllerModel() {
+			continue
+		}
+
+		// Now check for the validation of a model upgrade.
+		if err := m.validateSeriesUpgradeForSeriesUpgrade(st, args.Force); err != nil {
+			results.Results[i] = params.ErrorResult{Error: common.ServerError(err)}
+			continue
+		}
+	}
+	return results, nil
+}
+
+func (m *ModelManagerAPI) validateSeriesUpgradeForSeriesUpgrade(st State, force bool) error {
+	machines, err := st.AllMachines()
```

Comment:
It will be faster to add a new method to state for all the upgrade locks and just check that there are none.